lib/krb5/plugin.c

```diff
@@ -33,34 +35,165 @@
 
 #include "krb5_locl.h"
 
-#ifdef HAVE_DLFCN_H
-#include <dlfcn.h>
+/*
+ * Definitions:
+ *
+ *	module	    - a category of plugin module, identified by subsystem
+ *		      (typically "krb5")
+ *	dso	    - a library for a module containing a map of plugin
+ *		      types to plugins (e.g. "service_locator")
+ *	plugin	    - a set of callbacks and state that follows the
+ *		      common plugin module definition (version, init, fini)
+ *
+ * Obviously it would have been clearer to use the term "module" rather than
+ * "DSO" given there is an internal "DSO", but "module" was already taken...
+ *
+ *	modules := { module: dsos }
+ *	dsos := { path, dsohandle, plugins-by-name }
+ *	plugins-by-name := { plugin-name: [plug] }
+ *	plug := { ftable, ctx }
+ *
+ * Modules are also used to store the instance of libraries that can be
+ * declared as dependencies by plugins.
+ *
+ * Some existing plugin consumers outside libkrb5 use the "krb5" module
+ * namespace, but going forward the module should match the consumer library
+ * name (e.g. libhdb should use the "hdb" module rather than "krb5").
+ */
+
+/* global module use, use copy_modules() accessor to access */
+static heim_dict_t __modules;
+
+/* for updating __modules; NB retain/release is atomic and does not need lock */
+static HEIMDAL_MUTEX modules_mutex = HEIMDAL_MUTEX_INITIALIZER;
```

Comment:
N.B., this doesn't work.  There's a race condition between the read of the global and the taking of a reference to the object.  See https://github.com/nicowilliams/ctp

IIUC we already had this problem, so I might allow it for now and immediately do something about it.

Reply:
One thing we could do for now is always take the lock even to read (well, to retain the reference).

Reply:
Or, if we never replace the `__modules`, then just use a `heim_once` to initialize it.

Reply:
Ah, you only ever write `__modules` once.